File: python/Lab2/q1.py
'''Created on 14 Feb 2015
Grab a stick. Break it in two. Now randomly break another piece in two. Carry out
this process a total of n times.
Write a program that takes n as input and outputs the probability that a triangle can
be formed out of any three of the resulting pieces.

Does not work'''
from random import uniform
import time

def breaker(n):
    stick1 = 1.0
    stickdic = set()
    for nums in range(0,n):
        stick2 = 0
        while stick2==0:
            stick2 = uniform(0,stick1)
        stick1-=stick2
        stickdic.add(stick2)
    stickdic.add(stick1)
    return stickdic

a = time.clock()
times = 100000
count = 0
breaks = 2
for nums in range(0,times):
    sticks = breaker(breaks)
    for n in sticks:
        if n > 0.5:
            break
    else:
        count+=1
print "With %s breaks the chance of making a triangle is %s%%" % (breaks, count/float(times)*100)
print "%s sticks were broken in %sms" % (times, time.clock()-a)

File: C++/Lab2/Q1.cpp
/*
 * Q1.cpp

 *
 *  Created on: 4 Aug 2015
 *      Author: toby
 *  Grab a stick. Break it in two. Now randomly break another piece in two. Carry out
 *  this process a total of n times.
 *  Write a program that takes n as input and outputs the probability that a triangle can
 *  be formed out of any three of the resulting pieces.
 *  This solution takes a stick breaks it, takes a random stick breaks it etc etc
 *  Gets an answer of roughy 19-20% for 2 breaks
 */
#include <iostream>
#include <random>
#include <set>
#include <algorithm>

int trianglechecker(std::vector<double> & newsticks, int breaks){
	//checks to see if it makes a triangle
	for (auto it = newsticks.cbegin(); it != newsticks.cend()-2; it++){
		if ((*(it)+*(it+1))>(*(it+2)))return 1;
	}
	return 0;
}

int main(){
	const int times = 1000000;
	int count = 0;
	const int breaks = 2;
	double stick1 = 1.0;
	std::random_device rd;
	std::mt19937 mt(rd());

	for(int i =0;i<times;i++){
		std::set<double> breakset;
		//picks n number of random points to break at and puts them in a set.
		while(breakset.size()<breaks){
			std::uniform_real_distribution<double> dist(0,1);
			double stick2 = 0;
			while(stick2==0){
				stick2 = dist(mt);
			}
			breakset.insert(stick2);
		}
		std::vector<double> newsticks;
		double count1 = 0;
		//breaks the stick at set points into an array.
		for (auto it = breakset.cbegin(); it != breakset.cend(); it++){
			newsticks.push_back((*it) - count1);
			count1=(*it);
		}
		newsticks.push_back(1-count1);

		std::sort(newsticks.begin(), newsticks.end());
		count+=trianglechecker(newsticks, breaks);
	}
	std::cout << "With " << breaks << " breaks the chance of making a triangle is " << (count/(double)times)*100 << "%" << std::endl;
}
